core/server/common/src/main/java/alluxio/underfs/UfsManager.java

```diff
@@ -0,0 +1,164 @@
+/*
+ * The Alluxio Open Foundation licenses this work under the Apache License, version 2.0
+ * (the "License"). You may not use this work except in compliance with the License, which is
+ * available at www.apache.org/licenses/LICENSE-2.0
+ *
+ * This software is distributed on an "AS IS" basis, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND,
+ * either express or implied, as more fully set forth in the License.
+ *
+ * See the NOTICE file distributed with this work for information regarding copyright ownership.
+ */
+
+package alluxio.underfs;
+
+import alluxio.AlluxioURI;
+import alluxio.Configuration;
+import alluxio.PropertyKey;
+import alluxio.util.IdUtils;
+
+import com.google.common.base.Objects;
+import com.google.common.base.Preconditions;
+import com.google.common.io.Closer;
+
+import java.io.Closeable;
+import java.io.IOException;
+import java.util.Map;
+import java.util.concurrent.ConcurrentHashMap;
+
+import javax.annotation.concurrent.ThreadSafe;
+
+/**
+ * A class manages the UFS used by different worker services.
+ */
+@ThreadSafe
+public class UfsManager implements Closeable {
+
+  /**
+   * The key of the UFS cache.
+   */
+  private static class Key {
+    private final String mScheme;
+    private final String mAuthority;
+    private final Map<String, String> mProperties;
+
+    Key(AlluxioURI uri, Map<String, String> properties) {
+      mScheme = uri.getScheme() == null ? "" : uri.getScheme().toLowerCase();
+      mAuthority = uri.getAuthority() == null ? "" : uri.getAuthority().toLowerCase();
+      mProperties = (properties == null || properties.isEmpty()) ? null : properties;
+    }
+
+    @Override
+    public int hashCode() {
+      return Objects.hashCode(mScheme, mAuthority, mProperties);
+    }
+
+    @Override
+    public boolean equals(Object object) {
+      if (object == this) {
+        return true;
+      }
+
+      if (!(object instanceof Key)) {
+        return false;
+      }
+
+      Key that = (Key) object;
+      return Objects.equal(mAuthority, that.mAuthority) && Objects
+          .equal(mProperties, that.mProperties) && Objects.equal(mScheme, that.mScheme);
+    }
+
+    @Override
+    public String toString() {
+      return Objects.toStringHelper(this)
+          .add("authority", mAuthority)
+          .add("scheme", mScheme)
+          .add("properties", mProperties).toString();
+    }
+  }
+
+  /** Maps from key to {@link UnderFileSystem} instances. */
+  private final ConcurrentHashMap<Key, UnderFileSystem> mUnderFileSystemMap =
+      new ConcurrentHashMap<>();
+  /** Maps from mount id to {@link UnderFileSystem} instances. */
+  private final ConcurrentHashMap<Long, UnderFileSystem> mMountIdToUnderFileSystemMap =
```

Comment:
@gpang currently all ufses are indexed by mount id, so it is the same set of ufs instances in these two maps, one indexed by mount id, so we could quickly retrieve the ufs by id, one indexed by (scheme/authority/conf), so we know if we have cached this ufs instance before to reuse them when possible.